package/MDAnalysis/lib/distances.py

```diff
@@ -128,96 +128,184 @@ def _run(funcname, args=None, kwargs=None, backend="serial"):
+    boxtype = None
+    if box.shape == (6,):
+        if np.all(box[3:] == 90.):
+            boxtype = 'ortho'
+            checked_box = box[:3]
+        else:
+            boxtype = 'tri_vecs'
+            checked_box = triclinic_vectors(box)
+    elif box.shape == (3,):
         boxtype = 'ortho'
+        checked_box = box
     elif box.shape == (3, 3):
-        if np.all([box[0][1] == 0.0,  # Checks that tri box is properly formatted
-                      box[0][2] == 0.0,
-                      box[1][2] == 0.0]):
+        # Check if triclinic box vectors are improperly formatted:
+        if np.count_nonzero(np.triu(box, 1)):
+            checked_box = triclinic_vectors(triclinic_box(box[0], box[1],
+                                                          box[2]))
+        else:
+            checked_box = box
+        # Check if lower triangle is nonzero:
+        if np.count_nonzero(np.tril(checked_box, -1)):
             boxtype = 'tri_vecs'
+        # otherwise, the box is orthogonal and we return only the diagonal:
         else:
-            boxtype = 'tri_vecs_bad'
-    elif box.shape == (6,):
-        if np.all(box[3:] == 90.):
             boxtype = 'ortho'
-        else:
-            boxtype = 'tri_box'
+            checked_box = checked_box.diagonal().copy()
+    if boxtype is None:
+        raise ValueError("Box input not recognized, must be an array of box "
+                         "dimensions.")
+    elif not checked_box.flags['C_CONTIGUOUS']:
+        raise ValueError("Box must be a C-contiguous numpy array.")
+    return boxtype, checked_box
 
-    if boxtype == 'unknown':
-        raise ValueError("box input not recognised"
-                         ", must be an array of box dimensions")
 
-    return boxtype
+def _check_coord_array(coords, desc):
+    """Check if an array is a valid coordinate array.
 
+    The `coords` array must meet the following requirements:
+      * Must have a shape of ``(n, 3)``.
+      * Its dtype must be ``numpy.float32``.
 
-def _check_array(coords, desc):
-    """Check an array is a valid array of coordinates
+    Parameters
+    ----------
+    coords : numpy.ndarray
+        The coordinate array to check for validity.
+    desc : str
+        Name of the coordinate array. Only used in error messages.
 
-    Must be:
-       (n,3) in shape
-       float32 data
+    Raises
+    ------
+    ValueError
+        If `coords` has a wrong shape.
+    TypeError
+        If the dtype of `coords` is not ``numpy.float32``.
     """
-    if (coords.ndim != 2 or coords.shape[1] != 3):
-        raise ValueError("{0} must be a sequence of 3 dimensional coordinates"
-                         "".format(desc))
-    _check_array_dtype(coords, desc)
+    _check_coord_array_shape_2d(coords, desc)
+    if coords.dtype != np.float32:
+        raise TypeError("{0} must be of type numpy.float32".format(desc))
+# The following two lines would break a lot of tests. WHY?!
```

Comment:
What tests fail here?  Is it that some coords arrays are often non C contiguous?

Reply:
Ignore it, that's just a personal note I forgot to delete. Requiring coordinate arrays to be C-contiguous doesn't really make sense here since all the functions have calls to `np.astype(...)` which will return a continuous copy anyway.